### DISCUSSION TITLE
Custom plugins feature request

### DISCUSSION THREAD

**Discussion by JennToo** [renovatebot/renovate#22150] [Suggest an Idea]
**What would you like Renovate to be able to do?**
We have some internal packaging systems with lock files that we would like to be able to have automatically update with Renovate (Pro). 

**Describe the solution you'd like**
Since it is a proprietary format, some sort of custom plugin would be needed.

**Describe alternatives you've considered**
We have separate systems that update these lock files today, but would like to consolidate them into a single Renovate instance.

**Additional context**
Since the package format we're using isn't supported currently in Renovate, we would need to be able to add both a custom datasource as well as lock file.

We also would not want to open-source the plugin, if possible. In reality we probably could if this was a big requirement, but I don't think anyone else would find it useful.

**Comment by rarkins** [comment]
You may not need a custom datasource if your package manager is still installing existing things like npm packages, docker images, etc. You only need a custom datasource if you pull available versions from a custom API.

What type of versioning do you use? e.g. is it semver?

**Comment by JennToo** [comment]
It's semver with extras (we include variant names in the package versions as well)

Internally it's actually just a Debian-style repository. Where things get weird is how we install the packages, and the Debian ecosystem doesn't really have a concept of lockfiles so we made our own.

**Comment by rarkins** [comment]
The new regex versioning scheme might work out of the box for you (with appropriate configuration).

Sounds like you need a new datasource and manager, unless it’s a datasource you think could be useful to others. 

If we switch to a mode where we dynamically “discover” managers and datasources then plugins could be achieved by taking the standard docker images we distribute and adding files as a layer on top

**Comment by alexrashed** [comment]
@rarkins Are there any plans to do so?
We are currently facing the problem that we would love to use renovatebot for a pretty specific dependency: The OpenAPI specification for our backend services, which we use to generate all the client services. The generation is wrapped in an Angular builder, which is configured in the project's angular.json file.

Long story short: We would need a plugin mechanism to add a custom datasource and manager which checks for a new version of the OpenAPI spec and updates the version in the angular.json, respectively.

**Comment by rarkins** [comment]
@alexrashed right now this is unfortunately not something we are likely to implement any time soon due to a long list of high priority features we need to work on first, and this is pretty niche.

**Comment by alexrashed** [comment]
That's a pity but totally reasonable. Thanks for the quick response.
We'll probably refactor our generation s.t. the OpenAPI spec is defined as an NPM package.
Then we would be able to use the standard NPM features of your awesome bot.
Thanks for the great work!

**Comment by JennToo** [comment]
For our own part we've worked around this as well. We build a custom docker image already to fix some GHE API URL bugs (I really need to get that up in a PR at some point).

Since we're already building an image with some minor changes, we also just drop in a few specific features that we need. The renovate code is laid out well enough that merging in new changes from upstream is generally uneventful.

The hope was one day we would switch to the "Renovate Pro" scheduler for automatically running the update jobs, but TBH the home-grown scheduler we've built is working pretty well and is simple enough to maintain (and gives us a lot of flexibility to scale-out)

**Comment by valorl** [comment]
I think some kind of extension/plugin mechanism would add a lot to the power/flexibility of Renovate. 

Here's an example that's pretty generalizable: To install [Flux](https://fluxcd.io/) into your Kubernetes cluster, you may want to use their [terraform provider](https://registry.terraform.io/providers/fluxcd/flux), specifically the [`flux_install`](https://registry.terraform.io/providers/fluxcd/flux/latest/docs/data-sources/install) data source, which provides a [`version`](https://registry.terraform.io/providers/fluxcd/flux/latest/docs/data-sources/install#version) variable. It would be great to get a PR that bumps that variable whenever there's a new version of Flux available.

There are similar patterns in other Terraform modules/providers, and also outside of Terraform, as well as various internal tools. The value of this feature would essentially be the possibility to _reuse Renovate's various platform integrations for PRs_ where user provides the logic for looking up new versions and generate a code patch. 

As for the implementation, I'm thinking an "exec" approach would be the most flexible: User configures a command to run. When executed by Renovate (in the repository), the command is expected to produce a diff/patch, which renovate then creates a PR with. Since it's just a command with a STDOUT/file-based interface, it's up to the user and the complexity of the "bump" whether the logic should be written with coreutils/bash/python or a completely external binary.

**Comment by valorl** [comment]
On second thought :facepalm:... I think what I'm describing may be achieved using `regex` manager and e.g. the `github-releases` data source.
